lib/pbench/server/database/models/api_keys.py

```diff
@@ -0,0 +1,128 @@
+from typing import Optional
+
+from flask import current_app
+import jwt
+from sqlalchemy import Column, ForeignKey, String
+from sqlalchemy.orm import relationship
+
+from pbench.server.database.database import Database
+from pbench.server.database.models import decode_integrity_error, TZDateTime
+from pbench.server.database.models.users import User
+
+# Module private constants
+_TOKEN_ALG_INT = "HS256"
+
+
+class APIKeyError(Exception):
+    """A base class for errors reported by the APIKey class."""
+
+    def __init__(self, message):
+        self.message = message
+
+    def __str__(self):
+        return repr(self.message)
+
+
+class DuplicateApiKey(APIKeyError):
+    """Attempt to commit a duplicate unique value."""
+
+    def __init__(self, cause: str):
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Duplicate api_key: {self.cause}"
+
+
+class NullKey(APIKeyError):
+    """Attempt to commit an APIkey row with an empty required column."""
+
+    def __init__(self, cause: str):
+        self.cause = cause
+
+    def __str__(self) -> str:
+        return f"Missing required key: {self.cause}"
```

Comment:
I keep telling myself "maybe this is OK", but every time I see it something says "it's a missing *value* not a missing *key*". Hmm. Then again, it's not necessarily "missing", as in `user_id = None` isn't "missing" but would still trigger the integrity error. I can also tell myself this shouldn't matter much as we're defaulting the `created` column in all cases, which defaults to the current time, we already would have failed creation if we somehow *generated* a null API key, and certainly if we didn't find a user ID from `Auth`... and even if it somehow happens, this only translates to a journal message we're not directly reporting to the client.